File: problem-006.py
#! /usr/bin/env python3

# Sum square difference
# Problem 6

# The sum of the squares of the first ten natural numbers is,
# 12 + 22 + ... + 102 = 385

# The square of the sum of the first ten natural numbers is,
# (1 + 2 + ... + 10)2 = 552 = 3025

# Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.
# Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

def BruteForceSolution(N):
  sumOfSquares = sum(i*i for i in range(1,N+1))
  squareOfSum = pow(sum(range(1, N+1)), 2)
  diff = squareOfSum - sumOfSquares
  print("Brute force solution: ", diff)
  return diff

def MathSolution(N):
  """Use sum of squares formula and sum of integers formula for O(1) solution"""
  sumOfSquares = N * (N+1) * (2*N + 1) // 6
  squareOfSum = pow((N + 1) * N // 2, 2)
  diff = squareOfSum - sumOfSquares
  print("Math solution: ", diff)
  return diff

assert(BruteForceSolution(10) == 2640)
assert(MathSolution(10) == 2640)
BruteForceSolution(100)
MathSolution(100)

File: swift/Sources/Problem006.swift
// Sum square difference
// Problem 6

// The sum of the squares of the first ten natural numbers is,
// 1^2 + 2^2 + ... + 10^2 = 385

// The square of the sum of the first ten natural numbers is,
// (1 + 2 + ... + 10)^2 = 55^2 = 3025

// Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.
// Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

import Swiftest

func sumSquareDifference(n:Int) -> Int {
  // Use sum of squares formula and sum of integers formula for O(1) solution
  let sumOfSquares = n * (n+1) * (2*n + 1) / 6
  let sumOfN = ((n+1) * n) / 2
  let squareOfSum = sumOfN * sumOfN
  return squareOfSum - sumOfSquares
}

class Problem006Spec : Swiftest.Spec {
  let spec = describe("Problem 006 - Sum square difference") {
    it("should be 2640 for 1...10") {
      expect(sumSquareDifference(10)).to.equal(2640)
    }
    it("should be 25164150 for 1...100") {
      expect(sumSquareDifference(100)).to.equal(25164150)
    }
  }
}

